0000-encryption-api/README.md

```diff
@@ -0,0 +1,128 @@
+that could be implemented:
+
+- Aes256Gcm
+- Chacha20Poly1305
+- Twofish256CbcHmacSha2_256
+- Blowfish448HmacSha2_256
+
+Secret-key encryption methods generally require the use of *nonces* or *initialization vectors (iv)*, the two terms are often used interchangeably.
+We use the term *nonce* here. Nonces must be unique for a given key and message pair for certain encryption schemes to be prevent certain types of attacks.
+To simplify the interfaces for Ursa users, *nonce* generation will be handled by Ursa and not a required concern for developers.
+A lower level layer can expose this for more advanced crypto consumption like key homomorphic PRFs where nonce reuse might be desirable.
+
+Authenticated encryption methods are strongly preferred and any other methods should be considered lower level–only used by those who know what they are doing.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+All proposed code will be implemented in **libursa**.
+**Libursa** will use existing traits and structures where possible.
+`KeyGenOption`, `PublicKey`, `PrivateKey`, `SecretKey`, and `MacKey`
+are already implemented.
+
+Rust structs and traits will look and work like so:
+```rust
+pub trait PublicKeyEncryptionScheme {
+    fn new() -> Self;
+    fn keypair(&self, options: Option<KeyGenOption>) -> Result<(PublicKey, PrivateKey), CryptoError>;
+    fn encrypt(&self, plaintext: &[u8], pk: &PublicKey) -> Result<Vec<u8>, CryptoError>;
+    fn decrypt(&self, ciphertext: &[u8], sk: &PrivateKey) -> Result<Vec<u8>, CryptoError>;
+    fn private_key_size() -> usize;
+    fn public_key_size() -> usize;
+}
+
+pub trait SecretKeyEncryptionScheme {
+    fn new() -> Self;
+    fn genkey(&self, options: Option<KeyGenOption>) -> Result<SecretKey, CryptoError>;
+    fn encrypt(&self, plaintext: &[u8], sk: &SecretKey) -> Result<Vec<u8>, CryptoError>;
+    fn decrypt(&self, ciphertext: &[u8], sk: &SecretKey) -> Result<Vec<u8>, CryptoError>;
+    fn key_size() -> usize;
+    fn tag_size() -> usize;
+}
+```
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+Other libraries tie encryption methods to key objects instead of algorithm objects.
+The disadvantage of this model becomes apparent when multiple keys must be used for encryption
+and when safeguarding them.
+
+The key object model requires creating a new object then calling the encryption on the object.
+Our model considers keys as data blobs and only need one instance of an encryption scheme which allows any
+key to be passed without creating a new instance.
+
+Safeguarding should be easier where the key is a data blob that can be stored in secure locations.
+Key objects tend to encode more than just data when serialized.
+
+Other crypto libraries like libsodium and secp256k1 also follow this model with openssl beginning to adopt it as well.
+
+# Prior art
+[prior-art]: #prior-art
```

Comment:
I generally expect a prior art section to list previous work that is similar, along with brief descriptions of that work.
This section feels unnecessarily critical to me. I think it would be better to allow the reader to judge for themselves.
E.g. Openssl is a library that has many cryptographic algorithms but is written in "C"
might work better if changed to:
Openssl is a library with many cryptographic algorithms that is written in "C"

The same information is conveyed, but the opinion of the writer is not so blatant.